folding-schemes/src/folding/hypernova/decider_eth_circuit.rs

```diff
@@ -0,0 +1,630 @@
+
+        (computed_v).enforce_equal(&v)?;
+        Ok(())
+    }
+}
+
+/// Circuit that implements the in-circuit checks needed for the HyperNova's onchain (Ethereum's
+/// EVM) verification.
+#[derive(Clone, Debug)]
+pub struct DeciderEthCircuit<C1, GC1, C2, GC2, CS1, CS2, const H: bool = false>
+where
+    C1: CurveGroup,
+    GC1: CurveVar<C1, CF2<C1>>,
+    C2: CurveGroup,
+    GC2: CurveVar<C2, CF2<C2>>,
+    CS1: CommitmentScheme<C1, H>,
+    CS2: CommitmentScheme<C2, H>,
+{
+    _c1: PhantomData<C1>,
+    _gc1: PhantomData<GC1>,
+    _c2: PhantomData<C2>,
+    _gc2: PhantomData<GC2>,
+    _cs1: PhantomData<CS1>,
+    _cs2: PhantomData<CS2>,
+
+    /// E vector's length of the CycleFold instance witness
+    pub cf_E_len: usize,
+    /// CCS of the Augmented Function circuit
+    pub ccs: CCS<C1::ScalarField>,
+    /// R1CS of the CycleFold circuit
+    pub cf_r1cs: R1CS<C2::ScalarField>,
+    /// CycleFold PedersenParams over C2
+    pub cf_pedersen_params: PedersenParams<C2>,
+    pub poseidon_config: PoseidonConfig<CF1<C1>>,
+    /// public params hash
+    pub pp_hash: Option<C1::ScalarField>,
+    pub i: Option<CF1<C1>>,
+    /// initial state
+    pub z_0: Option<Vec<C1::ScalarField>>,
+    /// current i-th state
+    pub z_i: Option<Vec<C1::ScalarField>>,
+    /// Nova instances
+    pub U_i: Option<LCCCS<C1>>,
+    pub W_i: Option<Witness<C1::ScalarField>>,
+    pub u_i: Option<CCCS<C1>>,
+    pub w_i: Option<Witness<C1::ScalarField>>,
+    pub U_i1: Option<LCCCS<C1>>,
+    pub W_i1: Option<Witness<C1::ScalarField>>,
+    pub nimfs_proof: Option<NIMFSProof<C1>>,
+    // rho_0 is the first and only rho in the 'rho_powers' array, since it comes from NIMFS-folding
+    // only 2 instances.
+    pub rho_0: Option<C1::ScalarField>,
+    /// CycleFold running instance
+    pub cf_U_i: Option<NovaCommittedInstance<C2>>,
+    pub cf_W_i: Option<NovaWitness<C2>>,
+
+    /// KZG challenge & eval
+    pub kzg_challenge: Option<C1::ScalarField>,
+    pub eval_W: Option<C1::ScalarField>,
+}
+
+impl<C1, GC1, C2, GC2, CS1, CS2, const H: bool> DeciderEthCircuit<C1, GC1, C2, GC2, CS1, CS2, H>
+where
+    C1: CurveGroup,
+    C2: CurveGroup,
+    GC1: CurveVar<C1, CF2<C1>> + ToConstraintFieldGadget<CF2<C1>>,
+    GC2: CurveVar<C2, CF2<C2>> + ToConstraintFieldGadget<CF2<C2>>,
+    CS1: CommitmentScheme<C1, H>,
+    // enforce that the CS2 is Pedersen commitment scheme, since we're at Ethereum's EVM decider
+    CS2: CommitmentScheme<C2, H, ProverParams = PedersenParams<C2>>,
+    <C1 as Group>::ScalarField: Absorb,
+    <C1 as CurveGroup>::BaseField: PrimeField,
+{
+    pub fn from_hypernova<FC: FCircuit<C1::ScalarField>>(
+        hn: HyperNova<C1, GC1, C2, GC2, FC, CS1, CS2, H>,
+    ) -> Result<Self, Error> {
+        // compute the U_{i+1}, W_{i+1}, by folding the last running & incoming instances
+        let mut transcript = PoseidonSponge::<C1::ScalarField>::new(&hn.poseidon_config);
+        transcript.absorb(&hn.pp_hash);
```

Comment:
hmmm. Maybe we could check that this is indeed binding with someone no? In Halo2 I know we bind the Vkey (so circuit description) to the transcript at first. Do you mean that? Or when talk about parameters you talk about pedersen ones?

Reply:
Here parameters refers to the folding scheme parameters, which includes the curve parameters, the commitment scheme being used parameters, etc. In this way, for example, if you reuse all your witness and fold it but you change of curve, you will obtain a different hash (challenge), since you use a different set of parameters, and thus a different `pp_hash`.

If it helps, to have other examples, this is how it looks in microsoft/Nova: https://github.com/microsoft/Nova/blob/d2c52bd73e6a91c20f23ae4971f24ad70a9d0395/src/lib.rs#L56, and here how it is used in a hash: https://github.com/microsoft/Nova/blob/d2c52bd73e6a91c20f23ae4971f24ad70a9d0395/src/lib.rs#L303
similar idea in Nexus: https://github.com/nexus-xyz/nexus-zkvm/blob/364499635cd2cbb766a0e15f7190fd0a7edbffb7/nova/src/circuits/nova/public_params.rs#L60